File: Graph/Leetcode/Medium/CloneGraph/CloneGraph.java
// https://leetcode.com/problems/clone-graph/

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Node {

    public int val;
    public List<Node> neighbors;

    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }

    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }

    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}

public class CloneGraph {

    public Node cloneGraph(Node node) {
        // bfs
        if(node == null) return null;
        Node newNode = new Node(node.val);

        Deque<Node> que = new ArrayDeque<>();
        que.offer(node);
        Map<Node, Node> mp = new HashMap<>(); // og and copy
        mp.put(node, newNode);

        while(!que.isEmpty()) {
            Node currOg = que.poll(), currCopy = mp.get(currOg);
            for(Node neighbor: currOg.neighbors) {
                Node existingCopy = mp.getOrDefault(neighbor, null);
                if(existingCopy != null) {
                    currCopy.neighbors.add(existingCopy);
                } else {
                    Node newNeighbor = new Node(neighbor.val);
                    currCopy.neighbors.add(newNeighbor);
                    mp.put(neighbor, newNeighbor);
                    que.offer(neighbor);
                }
            }
        }

        return newNode;
    }
}

File: Graph/Leetcode/Medium/CloneGraph/CloneGraph.cpp
// https://leetcode.com/problems/clone-graph/

#include<bits/stdc++.h>
using namespace std;

class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};

class Solution
{
public:
    Node *cloneGraph(Node *node)
    {
        if (!node)
            return nullptr;
        Node *newNode = new Node(node->val);
        queue<Node *> que;
        unordered_map<Node *, Node *> mp;
        que.push(node);
        mp[node] = newNode;

        while (!que.empty())
        {
            Node *currOg = que.front();
            que.pop();
            Node *currCopy = mp[currOg];
            for (auto &neighbor : currOg->neighbors)
            {
                Node *existingNeighbor = mp[neighbor];
                if (existingNeighbor)
                {
                    currCopy->neighbors.push_back(existingNeighbor);
                }
                else
                {
                    Node *newNeighbor = new Node(neighbor->val);
                    currCopy->neighbors.push_back(newNeighbor);
                    que.push(neighbor);
                    mp[neighbor] = newNeighbor;
                }
            }
        }
        return newNode;
    }
};